tools/twix/src/panels/enum_plot.rs

```diff
@@ -74,272 +73,335 @@ impl Segment {
+    messages_count: usize,
+    last_error: Option<String>,
+}
+
+impl SegmentData {
+    fn subscribe(&mut self, nao: Arc<Nao>) {
+        self.change_buffer = match CyclerOutput::from_str(&self.output_key) {
+            Ok(output) => {
+                let buffer = nao.subscribe_changes(output);
+                self.last_error = None;
+                Some(buffer)
+            }
+            Err(error) => {
+                error!("Failed to subscribe: {:#}", error);
+                self.last_error = Some(error.to_string());
+                None
+            }
+        };
+    }
+
+    fn show_settings(&mut self, ui: &mut Ui, nao: Arc<Nao>) {
+        let subscription_field =
+            ui.add(CompletionEdit::outputs(&mut self.output_key, nao.as_ref()));
+
+        if subscription_field.changed() {
+            info!("Subscribing: {}", self.output_key);
+            if let Some(change_buffer) = self.change_buffer.as_mut() {
+                change_buffer.reset();
+            }
+            self.subscribe(nao);
+        }
+
+        if let Some(error) = self.last_error.as_ref() {
+            ui.colored_label(Color32::RED, error);
+        }
+    }
+
+    fn segments(&mut self) -> Vec<Segment> {
+        self.change_buffer
+            .as_ref()
+            .and_then(|change_buffer| match change_buffer.get_buffered() {
+                Ok(change_buffer_update) => {
+                    let mut segments = Vec::new();
+
+                    self.messages_count = change_buffer_update.message_count;
+
+                    for (start, end) in change_buffer_update.updates.iter().tuple_windows() {
+                        segments.push(Segment {
+                            start: start.message_number,
+                            end: end.message_number,
+                            value: start.value.clone(),
+                        });
+                    }
+
+                    if let Some(last_change) = change_buffer_update.updates.last() {
+                        segments.push(Segment {
+                            start: last_change.message_number,
+                            end: self.messages_count,
+                            value: last_change.value.clone(),
+                        });
+                    }
+
+                    Some(segments)
+                }
+                Err(error) => {
+                    self.last_error = Some(error);
+
+                    None
+                }
+            })
+            .unwrap_or_default()
+    }
+
+    fn clear(&mut self) {
+        if let Some(change_buffer) = self.change_buffer.as_ref() {
+            change_buffer.clear();
+        }
+        self.messages_count = 0;
+    }
+}
+
 pub struct EnumPlotPanel {
     nao: Arc<Nao>,
-    changes: Vec<Change>,
-    messages_count: usize,
-    scroll_position: f64,
-    viewport_width: f64,
-    change_buffer: Option<ChangeBuffer>,
-    output_key: String,
+    segment_datas: Vec<SegmentData>,
```

Comment:
"data" is already plural 😉 
```suggestion
    segment_data: Vec<SegmentData>,
```

Reply:
At least it's consistent with the plot panel. :upside_down_face: 
https://github.com/HULKs/hulk/blob/ca9cb0cedf97f59bc4e2eff577fa367e5061edee/tools/twix/src/panels/plot.rs#L189

Reply:
And that is also incorrect ;)